src/types/utils.cpp

```diff
@@ -631,6 +629,214 @@ bool Utils::IsValidHandle(const HANDLE handle) noexcept
+Utils::Pipe Utils::CreateOverlappedPipe(DWORD openMode, DWORD bufferSize)
+{
+    // Cache a handle to the pipe driver.
+    static const auto pipeDirectory = []() {
+        UNICODE_STRING path = RTL_CONSTANT_STRING(L"\\Device\\NamedPipe\\");
+
+        OBJECT_ATTRIBUTES objectAttributes{
+            .Length = sizeof(OBJECT_ATTRIBUTES),
+            .ObjectName = &path,
+        };
+
+        wil::unique_hfile dir;
+        IO_STATUS_BLOCK statusBlock;
+        THROW_IF_NTSTATUS_FAILED(NtCreateFile(
+            /* FileHandle        */ dir.addressof(),
+            /* DesiredAccess     */ SYNCHRONIZE | GENERIC_READ,
+            /* ObjectAttributes  */ &objectAttributes,
+            /* IoStatusBlock     */ &statusBlock,
+            /* AllocationSize    */ nullptr,
+            /* FileAttributes    */ 0,
+            /* ShareAccess       */ FILE_SHARE_READ | FILE_SHARE_WRITE,
+            /* CreateDisposition */ FILE_OPEN,
+            /* CreateOptions     */ FILE_SYNCHRONOUS_IO_NONALERT,
+            /* EaBuffer          */ nullptr,
+            /* EaLength          */ 0));
+
+        return dir;
+    }();
+
+    LARGE_INTEGER timeout = { .QuadPart = -10'0000'0000 }; // 1 second
+    UNICODE_STRING emptyPath{};
+    IO_STATUS_BLOCK statusBlock;
+    OBJECT_ATTRIBUTES objectAttributes{
+        .Length = sizeof(OBJECT_ATTRIBUTES),
+        .ObjectName = &emptyPath,
+        .Attributes = OBJ_CASE_INSENSITIVE,
+    };
+    DWORD desiredAccess = 0;
+    DWORD shareAccess = 0;
+
+    switch (openMode)
+    {
+    case PIPE_ACCESS_INBOUND:
+        desiredAccess = SYNCHRONIZE | GENERIC_READ | FILE_WRITE_ATTRIBUTES;
+        shareAccess = FILE_SHARE_WRITE;
+        break;
+    case PIPE_ACCESS_OUTBOUND:
+        desiredAccess = SYNCHRONIZE | GENERIC_WRITE | FILE_READ_ATTRIBUTES;
+        shareAccess = FILE_SHARE_READ;
+        break;
+    default:
+        desiredAccess = SYNCHRONIZE | GENERIC_READ | GENERIC_WRITE;
+        shareAccess = FILE_SHARE_READ | FILE_SHARE_WRITE;
+        break;
+    }
+
+    wil::unique_hfile server;
+    objectAttributes.RootDirectory = pipeDirectory.get();
+    THROW_IF_NTSTATUS_FAILED(NtCreateNamedPipeFile(
+        /* FileHandle        */ server.addressof(),
+        /* DesiredAccess     */ desiredAccess,
+        /* ObjectAttributes  */ &objectAttributes,
+        /* IoStatusBlock     */ &statusBlock,
+        /* ShareAccess       */ FILE_SHARE_READ | FILE_SHARE_WRITE,
+        /* CreateDisposition */ FILE_CREATE,
+        /* CreateOptions     */ 0, // would be FILE_SYNCHRONOUS_IO_NONALERT for a synchronous pipe
+        /* NamedPipeType     */ FILE_PIPE_BYTE_STREAM_TYPE,
+        /* ReadMode          */ FILE_PIPE_BYTE_STREAM_MODE,
+        /* CompletionMode    */ FILE_PIPE_QUEUE_OPERATION, // would be FILE_PIPE_COMPLETE_OPERATION for PIPE_NOWAIT
+        /* MaximumInstances  */ 1,
+        /* InboundQuota      */ bufferSize,
+        /* OutboundQuota     */ bufferSize,
+        /* DefaultTimeout    */ &timeout));

```

Comment:
`shareAccess` is set above, **not used** here, and overwritten below

Reply:
Fixed. Not 100% sure how to best structure this code that it looks good.